src/Microsoft.AspNetCore.Mvc.Core/Internal/ResponseCacheFilter.cs

```diff
@@ -110,6 +121,16 @@ public void OnActionExecuting(ActionExecutingContext context)
                 headers[HeaderNames.Vary] = VaryByHeader;
             }
 
+            if (VaryByQueryKeys != null)
+            {
+                var responseCacheFeature = context.HttpContext.Features.Get<IResponseCacheFeature>();
+                if (responseCacheFeature == null)
+                {
+                    throw new InvalidOperationException($"The response cache middleware must be added when using the {nameof(VaryByQueryKeys)} feature.");
+                }
+                responseCacheFeature.VaryByQueryKeys = VaryByQueryKeys;
```

Comment:
@pranavkm the server side caching is also called responsecaching. The reason I didn't separate it into a different attribute is that it may be confusing to have two different attributes and for people to remember which option lives where.